test/metabase/driver/common/parameters/dates_test.clj

```diff
@@ -61,7 +61,68 @@
            (params.dates/date-string->filter "next3days~" [:field "field" {:base-type :type/DateTime}]))))
   (testing "quarters (#21083)"
     (is (= [:time-interval [:field "field" {:base-type :type/DateTime}] -30 :quarter {:include-current false}]
-           (params.dates/date-string->filter "past30quarters" [:field "field" {:base-type :type/DateTime}])))))
+           (params.dates/date-string->filter "past30quarters" [:field "field" {:base-type :type/DateTime}]))))
+  (testing "relative (past) with starting from "
+    (is (= [:between
+            [:+ [:field "field" {:base-type :type/DateTime}] [:interval 3 :year]]
+            [:relative-datetime -3 :day]
+            [:relative-datetime 0 :day]]
+           (params.dates/date-string->filter "past3days-from-3years" [:field "field" {:base-type :type/DateTime}]))))
+  (testing "relative (next) with starting from"
+    (is (= [:between
+            [:+ [:field "field" {:base-type :type/DateTime}] [:interval -13 :month]]
+            [:relative-datetime 0 :hour]
+            [:relative-datetime 7 :hour]]
+           (params.dates/date-string->filter "next7hours-from-13months" [:field "field" {:base-type :type/DateTime}]))))
+  (testing "exclusions"
+    (t/with-clock (t/mock-clock #t "2016-06-07T12:13:55Z")
+      (testing "hours"
+        (is (= [:!=
+                [:field "field" {:base-type :type/DateTime, :temporal-unit :hour}]
+                "2016-06-07T00:00:00Z"]
+               (params.dates/date-string->filter "exclude-hours-0" [:field "field" {:base-type :type/DateTime}])))
+        (is (= [:!=
+                [:field "field" {:base-type :type/DateTime, :temporal-unit :hour}]
+                "2016-06-07T00:00:00Z"
+                "2016-06-07T23:00:00Z"]
+               (params.dates/date-string->filter "exclude-hours-0-23" [:field "field" {:base-type :type/DateTime}])))
+        (is (thrown? clojure.lang.ExceptionInfo #"Don't know how to parse date string \"exclude-hours-\""
+               (params.dates/date-string->filter "exclude-hours-" [:field "field" {:base-type :type/DateTime}])))
+        (is (thrown? clojure.lang.ExceptionInfo #"Don't know how to parse date string \"exclude-hours-24-3\""
+               (params.dates/date-string->filter "exclude-hours-24-3" [:field "field" {:base-type :type/DateTime}]))))
+      (testing "quarters"
+        (is (= [:!=
+                [:field "field" {:base-type :type/DateTime, :temporal-unit :quarter}]
```

Comment:
This doesn't work when testing with the frontend because this unit should be `:quarter-of-year`